src/scheduler.cpp

```diff
@@ -114,6 +114,27 @@ void CScheduler::scheduleFromNow(CScheduler::Function f, int64_t deltaMilliSecon
     schedule(f, boost::chrono::system_clock::now() + boost::chrono::milliseconds(deltaMilliSeconds));
 }
 
+void CScheduler::MockForward(boost::chrono::seconds delta_seconds)
+{
+    {
+        boost::unique_lock<boost::mutex> lock(newTaskMutex);
+
+        // use temp_queue to maintain updated schedule
+        std::multimap<boost::chrono::system_clock::time_point, Function> temp_queue;
+
+        for (const auto& element : taskQueue) {
+            boost::chrono::system_clock::time_point new_time = std::max(boost::chrono::system_clock::now() + delta_seconds, element.first) - delta_seconds;
```

Comment:
What's the problem of doing
```cpp
temp_queue.emplace_hint(temp_queue.cend(), element.first - delta_seconds, element.second);
```
Looks fine to me since `wait_until` returns when system time is later than that absolute time.

Didn't check the case for boost < 1.50.

Reply:
element.first - delta_seconds may underflow, so std::max prevents underflow.

Reply:
Underflow how? You mean be less than current time?

Reply:
if `element.first - delta_seconds < now` everything will process fine (as in the previous version you reviewed)

but if `delta_seconds > element.first` then there could be an underflow. seems highly unlikely because `element.first` is a time point, but doesn't hurt to have this guarantee

Reply:
This looks equivalent to

```
std::max(boost::chrono::system_clock::now(), element.first - delta_seconds);
```

Also perhaps you only want to invoke `now()` once and cache it rather than querying it for every queue element.

Reply:
@sipa that misses the point above. if delta_seconds > element.first, it underflows and then it can be greater than now().

I guess equivalents, if delta_seconds is too large, now() + delta_seconds may overflow... but that seems less likely because we don't know if this system_clock::now() which feeds into the queue is set to start at time 0 on startup.


Note also that caching now(), were emplace hint not used, would mean that the tmp queue would potentially be unstable ordered w.r.t. jobs with the same timeout.

Reply:
> if delta_seconds > element.first, it underflows and then it can be greater than now().

According to https://en.cppreference.com/w/cpp/chrono/duration, the representation type for the standard clock values is signed, so underflow should not be any issue (or at least not more than we need to worry about the clock overflowing in the first place).

> Note also that caching now(), were emplace hint not used, would mean that the tmp queue would potentially be unstable ordered w.r.t. jobs with the same timeout.

If that's a concern, it already is. There is no guarantee that subsequent calls to `now()` will be strictly increasing.

Reply:
caching `now()` seems good, I will update. 

RE: underflow / overflow- 
the intent is to ensure `delta_seconds` is reasonable. I'm thinking of asserting the value is between 0 & 3600 from both this `MockForward` method & the `mockscheduler` rpc method. 
and reverting to the original `temp_queue.emplace_hint(temp_queue.cend(), element.first - delta_seconds, element.second)`

Reply:
Ah -- note that it's boost and not std... I think that the seconds duration is guaranteed to be signed 64, but I couldn't find anything on system_clock's duration. Good catch on the non steady adjustable clock of system_clock, I had missed that. Seems like a general thing that our scheduler should be on a monotonic hardware clock rather than something that adjusting the system date could adjust (which would be another way to do this patch -- just fast forward the system time).

Reply:
Even with a monotonic clock there is no guarantee that subsequent requests to `now()` are *strictly* increasing (they'll be *monotonically* increasing, but that's not enough to guarantee a stable sorting order absent hints).

Reply:
reverted to original (`element.first - delta_seconds`), but added checks in `MockForward` and `mockscheduler` that `delta_seconds` is between 0 and 3600. 

which means we no longer invoke `now()` =P 

lmk if there are any further concerns

Reply:
lgtm